scipy/signal/tests/test_signaltools.py

```diff
@@ -31,72 +28,92 @@
                                       _group_poles)
 from scipy.signal._upfirdn import _upfirdn_modes
 from scipy._lib import _testutils
-from scipy._lib._util import ComplexWarning, np_long, np_ulong
+from scipy._lib._util import ComplexWarning
 
+from scipy._lib._array_api import (
+    xp_assert_close, xp_assert_equal, is_numpy, is_torch, array_namespace,
+    assert_array_almost_equal, assert_almost_equal,
+)
+from scipy.conftest import array_api_compatible
+skip_xp_backends = pytest.mark.skip_xp_backends
+pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends")]
 
-class _TestConvolve:
 
-    def test_basic(self):
-        a = [3, 4, 5, 6, 5, 4]
-        b = [1, 2, 3]
+@skip_xp_backends(cpu_only=True, exception=['cupy'])
```

Comment:
here and elsewhere

```suggestion
@skip_xp_backends(cpu_only=True, exceptions=['cupy'])
```

Reply:
Great spot, thanks! Fixed. 

Re piecemeal reviews, of course no review << piecemeal reviews << a full review :-). Given the size of this PR, it's not reasonable to expect the latter, so anything you can do under real life constraints is greatly appreciated! 

Given the availablility all around, I'd keep my several enhancements (dedupe/merge `_support_alternative_backends` machinery with ndimage, array api for more functions etc) to follow-ups anyhow.